File: UICollectionViewLayoutDemo/ProductCollectionViewCell/HWCollectionViewWaterfallLayout.m
//
//  UIView+FourCornerRadius.m
//  hw
//
//  Created by WeiHu on 2017/2/27.
//  Copyright © 2017年 dj0708. All rights reserved.
//


#import "HWCollectionViewWaterfallLayout.h"
#import "PJProductDecorationView.h"
#import <objc/runtime.h>

NSString *const HWCollectionElementKindSectionHeader = @"HWCollectionElementKindSectionHeader";
NSString *const HWCollectionElementKindSectionFooter = @"HWCollectionElementKindSectionFooter";
NSString *const HWCollectionElementKindSectionDecoration = @"HWCollectionElementKindSectionDecoration";

static NSString * const HWLayoutItemCellKind = @"HWLayoutItemCellKind";
static NSString * const HWLayoutHeaderKind = @"HWLayoutHeaderKind";
static NSString * const HWLayoutFooterKind = @"HWLayoutFooterKind";
static NSString * const HWLayoutDecorationKind = @"HWLayoutDecorationKind";


@implementation UICollectionView (HeaderViewAndFooterView)

- (UIView *)hw_collectionHeaderView
{
    return objc_getAssociatedObject(self, _cmd);
}

- (void)setHw_collectionHeaderView:(UIView *)hw_collectionHeaderView
{
    [self.hw_collectionHeaderView removeFromSuperview];
    objc_setAssociatedObject(self, @selector(hw_collectionHeaderView), hw_collectionHeaderView, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    [self addSubview:hw_collectionHeaderView];
    [self.collectionViewLayout invalidateLayout];
    
}
- (UIView *)hw_collectionFooterView
{
    return objc_getAssociatedObject(self, _cmd);
}

- (void)setHw_collectionFooterView:(UIView *)hw_collectionFooterView
{
    [self.hw_collectionFooterView removeFromSuperview];
    objc_setAssociatedObject(self, @selector(hw_collectionFooterView), hw_collectionFooterView, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    [self addSubview:hw_collectionFooterView];
    [self.collectionViewLayout invalidateLayout];
}

@end


@interface HWCollectionViewWaterfallLayout ()
@property (nonatomic, strong) NSDictionary *layoutInfo;

@end

@implementation HWCollectionViewWaterfallLayout

#pragma mark - Lifecycle

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self setup];
    }
    
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super init];
    if (self) {
        [self setup];
    }
    
    return self;
}

- (void)setup
{
    [self registerClass:[PJProductDecorationView class] forDecorationViewOfKind:HWCollectionElementKindSectionDecoration];
    
    self.itemInsets = UIEdgeInsetsZero;
    self.itemSize = CGSizeZero;
    self.headerHeight = 0;
    self.footerHeight = 0;
}

- (void)prepareLayout{
    [super prepareLayout];
    
    NSMutableDictionary *newLayoutInfo = [NSMutableDictionary dictionary];
    NSMutableDictionary *cellLayoutInfo = [NSMutableDictionary dictionary];
    NSMutableDictionary *headerLayoutInfo = [NSMutableDictionary dictionary];
    NSMutableDictionary *footerLayoutInfo = [NSMutableDictionary dictionary];
    NSMutableDictionary *decorationLayoutInfo = [NSMutableDictionary dictionary];
    
    NSInteger sectionCount = [self.collectionView numberOfSections];
    
    //CollectionHeaderView
    
    CGFloat top = 0.0;
    if (self.collectionView.hw_collectionHeaderView) {
        top += self.collectionView.hw_collectionHeaderView.bounds.size.height;
    }
    NSLog(@"%@",NSStringFromCGRect(self.collectionView.hw_collectionHeaderView.frame));
    
    for (NSInteger section = 0; section < sectionCount; section++) {
        NSInteger itemCount = [self.collectionView numberOfItemsInSection:section];
        
        //Header
        NSIndexPath *headerIndexPath = [NSIndexPath indexPathForItem:0 inSection:section];
        UICollectionViewLayoutAttributes *headerAttributes = [UICollectionViewLayoutAttributes layoutAttributesForSupplementaryViewOfKind:HWCollectionElementKindSectionHeader withIndexPath:headerIndexPath];
        CGFloat headerHeight = self.headerHeight;
        if ([self.layoutDelegate respondsToSelector:@selector(collectionView:layout:referenceSizeForHeaderInSection:)]) {
            headerHeight = [self.layoutDelegate collectionView:self.collectionView layout:self referenceSizeForHeaderInSection:section];
        }
        //header insets
        
        headerAttributes.frame = CGRectMake(0, top, CGRectGetWidth(self.collectionView.frame), headerHeight);
        headerLayoutInfo[headerIndexPath] = headerAttributes;
        top += headerHeight;
        
        //item set up
        CGFloat itemHeight = 0.0;
        for (NSInteger item = 0; item < itemCount; item++) {
            
            NSIndexPath *indexPath = [NSIndexPath indexPathForItem:item inSection:section];
            UICollectionViewLayoutAttributes *itemAttributes =
            [UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:indexPath];
            
            //itemSize
            CGSize itemSize = self.itemSize;
            if ([self.layoutDelegate respondsToSelector:@selector(collectionView:layout:sizeForItemAtIndexPath:)]) {
                itemSize = [self.layoutDelegate collectionView:self.collectionView layout:self sizeForItemAtIndexPath:indexPath];
            }
            //itemInsets
            UIEdgeInsets itemInsets = self.itemInsets;
            if ([self.layoutDelegate respondsToSelector:@selector(collectionView:layout:insetForItemAtIndexPath:)]) {
                itemInsets = [self.layoutDelegate collectionView:self.collectionView layout:self insetForItemAtIndexPath:indexPath];
            }
            
            //item frame
            CGFloat item_x = itemInsets.left;
            CGFloat item_y = top + itemHeight + itemInsets.top;
            CGFloat item_width = itemSize.width - itemInsets.left - itemInsets.right;
            CGFloat item_height = itemSize.height - itemInsets.top - itemInsets.bottom;
            
            itemAttributes.frame = CGRectMake(item_x,item_y, item_width, item_height);
            cellLayoutInfo[indexPath] = itemAttributes;
            
            itemHeight += (itemSize.height);
        }
        
        top += itemHeight;
        
        //Footer frame
        NSIndexPath *footerIndexPath = [NSIndexPath indexPathForItem:itemCount - 1 inSection:section];
        UICollectionViewLayoutAttributes *footerAttributes = [UICollectionViewLayoutAttributes layoutAttributesForSupplementaryViewOfKind:HWCollectionElementKindSectionFooter withIndexPath:footerIndexPath];
        
        CGFloat footerHeight = self.footerHeight;
        if ([self.layoutDelegate respondsToSelector:@selector(collectionView:layout:referenceSizeForHeaderInSection:)]) {
            footerHeight = [self.layoutDelegate collectionView:self.collectionView layout:self referenceSizeForFooterInSection:section];
        }
        
        footerAttributes.frame = CGRectMake(0, top, CGRectGetWidth(self.collectionView.frame), footerHeight);
        footerLayoutInfo[footerIndexPath] = footerAttributes;
        
        top += footerHeight;
        
        UIEdgeInsets decorationInsets = self.decorationInsets;
        if ([self.layoutDelegate respondsToSelector:@selector(collectionView:layout:insetForDecorationSectionAtIndex:)]) {
            decorationInsets = [self.layoutDelegate collectionView:self.collectionView layout:self insetForDecorationSectionAtIndex:section];
        }
        
        //Decoration frame
        NSIndexPath *decorationIndexPath = [NSIndexPath indexPathForItem:0 inSection:section];
        UICollectionViewLayoutAttributes *decorationAttributes = [UICollectionViewLayoutAttributes layoutAttributesForDecorationViewOfKind:HWCollectionElementKindSectionDecoration withIndexPath:decorationIndexPath];
        CGFloat decorationHeight = CGRectGetMaxY(footerAttributes.frame) - CGRectGetMinY(headerAttributes.frame);
        
        decorationAttributes.frame = CGRectMake(decorationInsets.left, CGRectGetMinY(headerAttributes.frame) + decorationInsets.top,CGRectGetWidth(self.collectionView.frame) - decorationInsets.left - decorationInsets.right, decorationHeight - decorationInsets.top - decorationInsets.bottom);
        decorationAttributes.zIndex = -1;
        decorationLayoutInfo[decorationIndexPath] = decorationAttributes;
        
    }
    
    //CollectionFooterView
    if (self.collectionView.hw_collectionFooterView) {
        self.collectionView.hw_collectionFooterView.frame = CGRectMake(self.collectionView.hw_collectionFooterView.frame.origin.x, top , self.collectionView.hw_collectionFooterView.frame.size.width, self.collectionView.hw_collectionFooterView.frame.size.height)
        ;
        top += self.collectionView.hw_collectionFooterView.bounds.size.height;
    }
    
    newLayoutInfo[HWLayoutItemCellKind] = cellLayoutInfo;
    newLayoutInfo[HWLayoutHeaderKind] = headerLayoutInfo;
    newLayoutInfo[HWLayoutFooterKind] = footerLayoutInfo;
    newLayoutInfo[HWLayoutDecorationKind] = decorationLayoutInfo;
    
    self.layoutInfo = newLayoutInfo;
}


- (NSArray *)layoutAttributesForElementsInRect:(CGRect)rect
{
    NSMutableArray *allAttributes = [NSMutableArray arrayWithCapacity:self.layoutInfo.count];
    
    [self.layoutInfo enumerateKeysAndObjectsUsingBlock:^(NSString *elementIdentifier,
                                                         NSDictionary *elementsInfo,
                                                         BOOL *stop) {
        [elementsInfo enumerateKeysAndObjectsUsingBlock:^(NSIndexPath *indexPath,
                                                          UICollectionViewLayoutAttributes *attributes,
                                                          BOOL *innerStop) {
            if (CGRectIntersectsRect(rect, attributes.frame)) {
                [allAttributes addObject:attributes];
            }
        }];
    }];
    
    return allAttributes;
}

- (CGSize)collectionViewContentSize
{
    NSInteger sectionCount = [self.collectionView numberOfSections];
    if (sectionCount <= 0) return [super collectionViewContentSize];
    NSInteger numberOfItemsInSection = [self.collectionView numberOfItemsInSection:sectionCount - 1];
    
    UICollectionViewLayoutAttributes *footAttributes = nil;
    NSDictionary *dic = self.layoutInfo[HWLayoutFooterKind];
    if (dic.count > 0) {
        footAttributes = self.layoutInfo[HWLayoutFooterKind][[NSIndexPath indexPathForRow:numberOfItemsInSection - 1 inSection:sectionCount - 1]];
    }else{
        footAttributes = self.layoutInfo[HWLayoutItemCellKind][[NSIndexPath indexPathForRow:numberOfItemsInSection - 1 inSection:sectionCount - 1]];
        
    }
    CGFloat height = CGRectGetMaxY(footAttributes.frame);
    if (self.collectionView.hw_collectionFooterView) {
        height += CGRectGetHeight(self.collectionView.hw_collectionFooterView.frame);
    }
    
    return CGSizeMake(self.collectionView.bounds.size.width, height);
}
@end

File: SwiftWaterfallLayout/WaterfallLayout/HWCollectionViewWaterfallLayout.swift
//
//  HWCollectionViewWaterfallLayout.swift
//  UICollectionViewLayoutDemoSwift
//
//  Created by WeiHu on 2017/4/1.
//  Copyright © 2017年 WeiHu. All rights reserved.
//

import UIKit

struct HWCollectionElementKindSection {
    static let header: String = "HWCollectionElementKindSectionHeader"
    static let footer: String = "HWCollectionElementKindSectionFooter"
    static let decoration: String = "HWCollectionElementKindSectionDecoration"
}

struct HWLayoutKind {
    static let cell: String = "HWLayoutKind.cell"
    static let header: String = "HWLayoutKind.header"
    static let footer: String = "HWLayoutKind.footer"
    static let decoration: String = "HWLayoutKind.decoration"
}

@objc protocol HWCollectionViewWaterfallLayoutDelegate: class {
    //item size
    @objc optional func collectionView(_ collectionView: UICollectionView, layout: HWCollectionViewWaterfallLayout,sizeForItemAtIndexPath: IndexPath) -> CGSize
    //Footer Insets
    @objc optional func collectionView(_ collectionView: UICollectionView, layout: HWCollectionViewWaterfallLayout,insetForFooterSectionAtIndex: IndexPath) -> UIEdgeInsets
    //Header Insets
    @objc optional func collectionView(_ collectionView: UICollectionView, layout: HWCollectionViewWaterfallLayout,insetForHeaderSectionAtIndex: IndexPath) -> UIEdgeInsets
    //item Insets
    @objc optional func collectionView(_ collectionView: UICollectionView, layout: HWCollectionViewWaterfallLayout,insetForItemAtIndexPath: IndexPath) -> UIEdgeInsets
    //decoration Insets
    @objc optional func collectionView(_ collectionView: UICollectionView, layout: HWCollectionViewWaterfallLayout,insetForDecorationSectionAtIndex: IndexPath) -> UIEdgeInsets
    //header height
    @objc optional func collectionView(_ collectionView: UICollectionView, layout: HWCollectionViewWaterfallLayout,referenceSizeForHeaderInSection: IndexPath) -> Float
    //footer height
    @objc optional func collectionView(_ collectionView: UICollectionView, layout: HWCollectionViewWaterfallLayout,referenceSizeForFooterInSection: IndexPath) -> Float
}
//Propertys
class HWCollectionViewWaterfallLayout: UICollectionViewLayout {
    var itemInsets: UIEdgeInsets = .zero
    var headerInsets: UIEdgeInsets = .zero
    var footerInsets: UIEdgeInsets = .zero
    var decorationInsets: UIEdgeInsets = .zero
    private var layoutInfo: [String : [IndexPath: UICollectionViewLayoutAttributes]]?
    
    var itemSize: CGSize = .zero
    var headerHeight: Float = 0.0
    var footerHeight: Float = 0.0
    weak var layoutDelegate: HWCollectionViewWaterfallLayoutDelegate?
    override init() {
        super.init()
        self.register(HWCollectionReusableDecorationView.self, forDecorationViewOfKind: HWCollectionElementKindSection.decoration)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    override func prepare() {
        
        var newLayoutInfo: [String : [IndexPath: UICollectionViewLayoutAttributes]] = [String : [IndexPath: UICollectionViewLayoutAttributes]]()
        var cellLayoutInfo: [IndexPath: UICollectionViewLayoutAttributes] = [IndexPath: UICollectionViewLayoutAttributes]()
        var headerLayoutInfo: [IndexPath: UICollectionViewLayoutAttributes] = [IndexPath: UICollectionViewLayoutAttributes]()
        var footerLayoutInfo: [IndexPath: UICollectionViewLayoutAttributes] = [IndexPath: UICollectionViewLayoutAttributes]()
        var decorationLayoutInfo: [IndexPath: UICollectionViewLayoutAttributes] = [IndexPath: UICollectionViewLayoutAttributes]()
        
        guard let collectionView = self.collectionView else { return }
        let sectionCount = collectionView.numberOfSections
        var top: Float = 0.0;
        //CollectionHeaderView
        if let hw_collectionHeaderView = collectionView.hw_collectionHeaderView{
            top += Float(hw_collectionHeaderView.frame.height)
        }
    
        for section in 0..<sectionCount {
            //Header
            
            let headerIndexPath: IndexPath = IndexPath(row: 0,section: section)
           
            let headerAttributes = UICollectionViewLayoutAttributes(forSupplementaryViewOfKind: HWCollectionElementKindSection.header, with: headerIndexPath)
            var headerHeight: Float = self.headerHeight;
            
            if let layoutDelegate = self.layoutDelegate, let method = layoutDelegate.collectionView?(collectionView, layout: self, referenceSizeForHeaderInSection: headerIndexPath){
               headerHeight = method
            }
            headerAttributes.frame = CGRect(x: 0, y: CGFloat(top), width: collectionView.frame.width,height: CGFloat(headerHeight));
            headerLayoutInfo[headerIndexPath] = headerAttributes;
            top += headerHeight;
            
            //item set up
            var itemHeight: Float = 0.0
            let itemCount: Int = collectionView.numberOfItems(inSection: section)
            for row in 0..<itemCount {
                let indexPath: IndexPath = IndexPath(row: row,section: section)
                let itemAttributes = UICollectionViewLayoutAttributes(forCellWith: indexPath)
        
                //itemSize
                var itemSize: CGSize = self.itemSize;
                if let layoutDelegate = self.layoutDelegate, let method = layoutDelegate.collectionView?(collectionView, layout: self, sizeForItemAtIndexPath: indexPath){
                    itemSize = method
                }
                
                //itemInsets
                var itemInsets: UIEdgeInsets = self.itemInsets;
                if let layoutDelegate = self.layoutDelegate, let method = layoutDelegate.collectionView?(collectionView, layout: self, insetForItemAtIndexPath: indexPath){
                    itemInsets = method
                }
                
                //item frame
                let item_x: Float = Float(itemInsets.left)
                let item_y: Float = top + itemHeight + Float(itemInsets.top)
                let item_width: Float = Float(itemSize.width - itemInsets.left - itemInsets.right)
                let item_height: Float = Float(itemSize.height - itemInsets.top - itemInsets.bottom)
                
                itemAttributes.frame = CGRect(x: CGFloat(item_x),y: CGFloat(item_y),width: CGFloat(item_width),height: CGFloat(item_height));
                cellLayoutInfo[indexPath] = itemAttributes
                
                itemHeight += Float(itemSize.height)
            }
             top += itemHeight
         
            //Footer frame
            let footerIndexPath = IndexPath(row: itemCount - 1, section: section)
            let footerAttributes = UICollectionViewLayoutAttributes(forSupplementaryViewOfKind: HWCollectionElementKindSection.footer,with: footerIndexPath)
            var footerHeight: Float = self.footerHeight
            
            if let layoutDelegate = self.layoutDelegate, let method = layoutDelegate.collectionView?(collectionView, layout: self, referenceSizeForFooterInSection: footerIndexPath){
                footerHeight = method
            }
            
            footerAttributes.frame = CGRect(x: CGFloat(0),y: CGFloat(top),width: collectionView.frame.width,height: CGFloat(footerHeight));
            footerLayoutInfo[footerIndexPath] = footerAttributes;
            top += footerHeight;
            
            //Decoration frame
            var decorationInsets = self.decorationInsets;
            if let layoutDelegate = self.layoutDelegate, let method = layoutDelegate.collectionView?(collectionView, layout: self, insetForDecorationSectionAtIndex: footerIndexPath){
                decorationInsets = method
            }
            let decorationIndexPath: IndexPath = IndexPath(row:0, section: section)
            
            let decorationAttributes: UICollectionViewLayoutAttributes = UICollectionViewLayoutAttributes(forDecorationViewOfKind: HWCollectionElementKindSection.decoration, with: decorationIndexPath)
            
            let decorationHeight = footerAttributes.frame.maxY - headerAttributes.frame.minY;
            
            decorationAttributes.frame = CGRect(x: decorationInsets.left, y: headerAttributes.frame.minY + decorationInsets.top,width: collectionView.frame.width - decorationInsets.left - decorationInsets.right, height: decorationHeight - decorationInsets.top - decorationInsets.bottom);
            decorationAttributes.zIndex = -1;
            decorationLayoutInfo[decorationIndexPath] = decorationAttributes;

            
        }
        
        //CollectionFooterView
        if let hw_collectionFooterView = collectionView.hw_collectionFooterView {
            hw_collectionFooterView.frame = CGRect(x: hw_collectionFooterView.frame.origin.x,y: CGFloat(top) ,width: hw_collectionFooterView.frame.size.width,height: hw_collectionFooterView.frame.size.height)
            ;
            top += Float(hw_collectionFooterView.bounds.height)
        }
        
        newLayoutInfo[HWLayoutKind.cell] = cellLayoutInfo;
        newLayoutInfo[HWLayoutKind.header] = headerLayoutInfo;
        newLayoutInfo[HWLayoutKind.footer] = footerLayoutInfo;
        newLayoutInfo[HWLayoutKind.decoration] = decorationLayoutInfo;
        
        self.layoutInfo = newLayoutInfo;
    }
    
    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        guard let layoutInfo = self.self.layoutInfo else { return super.layoutAttributesForElements(in: rect)}
        
        var allAttributes = [UICollectionViewLayoutAttributes]()

        layoutInfo.forEach { (_, elementsInfo) in
            
            elementsInfo.forEach({ (_,attributes) in
                if rect.intersects(attributes.frame) {
                    allAttributes.append(attributes)
                }
            })
            
        }
        return allAttributes;
    }
    override var collectionViewContentSize: CGSize{
    
        guard let collectionView = self.collectionView else { return super.collectionViewContentSize}
        let sectionCount: Int = collectionView.numberOfSections
        if (sectionCount <= 0) { return super.collectionViewContentSize}
        let numberOfItemsInSection: Int = collectionView.numberOfItems(inSection: sectionCount - 1)
    
        guard let layoutInfo = self.self.layoutInfo else { return super.collectionViewContentSize}
        let dic = layoutInfo[HWLayoutKind.footer]
        
        let lastAttributes: UICollectionViewLayoutAttributes = dic![IndexPath(row: numberOfItemsInSection - 1, section: sectionCount - 1)]!
        
        var height = lastAttributes.frame.maxY;
        if let hw_collectionFooterView = collectionView.hw_collectionFooterView {
            height += hw_collectionFooterView.frame.height;
        }
        return CGSize(width: collectionView.bounds.size.width,height: height);
        
        
    }
    
}

extension UICollectionView{
    private struct AssociatedKeys {
        static var header = "UICollectionView.Header"
        static var footer = "UICollectionView.Footer"
    }
    
    var hw_collectionHeaderView: UIView? {
        get {
            return objc_getAssociatedObject(self, &AssociatedKeys.header) as? UIView
        }
        set (headerView) {
            guard let headerView = headerView else { return }
            self.hw_collectionHeaderView?.removeFromSuperview()
            objc_setAssociatedObject(self, &AssociatedKeys.header, headerView, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
            self.addSubview(headerView)
        }
    }
    var hw_collectionFooterView: UIView? {
        get {
            return objc_getAssociatedObject(self, &AssociatedKeys.footer) as? UIView
        }
        set (footerView) {
            guard let footerView = footerView else { return }
            self.hw_collectionFooterView?.removeFromSuperview()
            objc_setAssociatedObject(self, &AssociatedKeys.footer, footerView, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
            self.addSubview(footerView)
        }
    }
}
